src/Common/src/Interop/FreeBSD/Interop.libutil.cs

```diff
@@ -0,0 +1,429 @@
+            int ki_flag2;       /* P2_* flags */
+            int ki_fibnum;      /* Default FIB number */
+            uint   ki_cr_flags;        /* Credential flags */
+            int ki_jid;         /* Process jail ID */
+            public int ki_numthreads;      /* XXXKSE number of threads in total */
+            public int ki_tid;         /* XXXKSE thread id */
+            fixed byte ki_pri[4];    /* process priority */
+            public rusage ki_rusage;   /* process rusage statistics */
+            /* XXX - most fields in ki_rusage_ch are not (yet) filled in */
+            rusage ki_rusage_ch;    /* rusage of children processes */
+            void *ki_pcb;        /* kernel virtual addr of pcb */
+            void    *ki_kstack;     /* kernel virtual addr of stack */
+            void    *ki_udata;      /* User convenience pointer */
+            public void *ki_tdaddr;  /* address of thread */
+
+            fixed long ki_spareptrs[KI_NSPARE_PTR];   /* spare room for growth */
+            fixed long    ki_sparelongs[KI_NSPARE_LONG];  /* spare room for growth */
+            long    ki_sflag;       /* PS_* flags */
+            public long    ki_tdflags;     /* XXXKSE kthread flag */
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        internal unsafe struct sysctl_name
+        {
+            public fixed int name[4];
+        }
+
+        [DllImport("libc", SetLastError = true)]
+        private static extern unsafe int sysctl(
+            int [] name,
+            int namelen,
+            byte* oldp,
+           ulong *oldlenp,
+           void *newp,
+           ulong newlen);
+        
+        /// <summary>
+        /// Queries the OS for the list of all running processes and returns the PID for each
+        /// </summary>
+        /// <returns>Returns a list of PIDs corresponding to all running processes</returns>
+        internal static unsafe int[] proc_listallpids()
+        {
+            int numProcesses=0;
+            int[] pids;
+
+            kinfo_proc * entries = getProcInfo(0, false, out numProcesses);
+
+            if (entries == null || numProcesses <= 0)
+            {
+                throw new Win32Exception(SR.CantGetAllPids);
+            }
+
+            Span<kinfo_proc>  list = new Span<kinfo_proc>(entries, numProcesses);
+            pids = new int[numProcesses];
+            int idx=0;
+            // walk thorough process list and skip kernel threads
+            for(int i=0; i < list.Length; i++) {
+                if (list[i].ki_ppid == 0)
+                {
+                    // skip kernel threads
+                    numProcesses-=1;
+                }
+                else
+                {
+                    pids[idx] = list[i].ki_pid;
+                    idx += 1;
+                }
+            }
+            // Remove extra elements
+            Array.Resize<int>(ref pids, numProcesses);
+            Marshal.FreeHGlobal((IntPtr)entries);
```

Comment:
This free should be done in a finally so we don't leak in the cases of exceptions getting thrown.